### PR TITLE
Allow table projection from within likeifyable SQL-compiled regular expressions

### PR DESCRIPTION
**Pull request by SimonJF**
# Issue

(Related in part to #754).

Previously, compiled regular expressions failed at runtime. As an example from GtoPdb:

```
for (strI <-- Tables.structural_info)
  where (strI.object_id == objectId && strI.species_id == 2)
  for (al <-- Tables.allele)
    where (strI.official_gene_id <> "" && al.accessions =~ /.*{strI.official_gene_id}.*/)
```

would fail due to the clause

```
al.accessions =~ /.*{strI.official_gene_id}.*/
```

within the `where` clause, since we're including a regular expression which projects from the `strI` table.

# Scope

This patch rectifies the above problem for `LIKE`-translatable queries (but not yet for arbitrary `RLIKE` queries).

# Proposal

To fix this, we need to do two things: generalise the return type of `likeify` to include SQL which allows strings, concatenation, and projection operators; and ensure the nested query translations occur inside regular expressions.

### CODE REVIEW

**Comment by jamescheney** [comment]
Huh.  While it's nice to be able to support this, I admit to being a little surprised it is even possible - I guess because SQL LIKE does its regular expression compiling as a run-time "eval"-like operation which is late enough that we can do other SQL string operations first.

I am wondering whether this same limitation also applies to flat queries - did we already support regexp antiquotation referring to query-generator-bound variables in flat queries, or would it be good to do the same thing to the flat query engine so that the same support is available tehre?

**Comment by SimonJF** [comment]
> Huh. While it's nice to be able to support this, I admit to being a little surprised it is even possible - I guess because SQL LIKE does its regular expression compiling as a run-time "eval"-like operation which is late enough that we can do other SQL string operations first.

Yes, it's pretty neat -- the nested query test case gets compiled as:

```
select (2) as "1@1",(t2345."2") as "1@2",(t2341."c") as "2"
from (select (t2340."name") as "1@1@name",(row_number() over (order by t2340."name")) as "2"
from "marketers" as t2340) as t2345,"mc" as t2341
where ((t2341."m") LIKE ('%' || t2345."1@1@name" || '%')) and ((t2341."m") = ('a'))
```

noting `LIKE ('%' || t2345."1@1@name" || '%')`.

> I am wondering whether this same limitation also applies to flat queries - did we already support regexp antiquotation referring to query-generator-bound variables in flat queries, or would it be good to do the same thing to the flat query engine so that the same support is available tehre?

My hunch is that the only thing stopping the flat evaluator from working was the `likeify` return type limitation: the edits to the nested query evaluator were only necessary in order to ensure that we were projecting from the correct table after shredding. That said, I'll add in a test.

**Comment by SimonJF** [comment]
Sorry, CI and PRs are being funky. The first run passed with the nested query. I've just updated with the flat query test (which passes locally) so hopefully that should go through, too.

**Comment by dhil** [comment]
> Sorry, CI and PRs are being funky. The first run passed with the nested query. I've just updated with the flat query test (which passes locally) so hopefully that should go through, too.

I think Travis CI missed you!

**Comment by SimonJF** [comment]
:upside_down_face: 

Seems like the build is queued at least, hopefully if it goes through, it'll trigger the necessary hooks...

https://travis-ci.org/github/links-lang/links/builds/742710312

**Comment by SimonJF** [comment]
Of course MySQL does things differently 🙄

**Comment by SimonJF** [comment]
Putting this on hold until #937 is done (thanks, MySQL)

**Comment by SimonJF** [comment]
Woo! OK, after that adventure, I think this is finally ready for review. 

As a pleasant side-effect, I also fixed string concatenation in MySQL more generally.

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:378]: Is there a reason for making li be a function of type `unit -> QL.t -> QL.t` rather than just leaving out the unit?  I guess this avoids nontermination but eta-expanding the RHS by the argument should also do it, and then you'd not have to write `li () xyzzy` everywhere.

```suggestion
    let li e = lins_inner (z, z_fields) ys e in
```

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:396]:
```suggestion
          Apply (Primitive "tilde", [li s; li r])
```

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:398]:
```suggestion
        Apply (Primitive f, List.map li es)
```

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:400]:
```suggestion
        Record (StringMap.map li fields)
```

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:408]:
```suggestion
          Variant ("Simply", li x)
```

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:410]:
```suggestion
          Variant ("Seq", Singleton (li r))
```

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:414]:
```suggestion
              function | Singleton x -> Singleton (li x) | _ -> assert false) rs))
```

File: core/query/evalNestedQuery.ml
Comment (jamescheney) [inline core/query/evalNestedQuery.ml:416]:
```suggestion
          Variant ("Quote", Variant ("Simply", li v))
```

File: core/query/sql.ml
Comment (jamescheney) [inline core/query/sql.ml:29]: I'm curious what is the benefit of having a separate case for LIKE strings as opposed to just using the string constant, ordinary projection, and string concatenation operators.  Clearly doing it this way makes it easier to recognize where match expressions have been translated to (LIKE) strings since regexp compilation is the only place where we do that.  However, what would break (or be unpleasantly harder) if we just compiled regexps to the existing Constant(String -), Project and Append operations?

**Review by jamescheney — CHANGES_REQUESTED** [review]
Looks reasonable.  I have a few nitpicky questions - I'm not insisting on changes to the code but if not it would be good to document the reasoning with comments in a few places.

**Comment by SimonJF** [comment]
Thanks for the excellent suggestions, @jamescheney! I've removed thunking in favour of eta-expansion (good trick), and removed the `SqlLike` construct in favour of compiling directly to the query language.

**Review by jamescheney — COMMENTED** [review]
A few tiny nitpicks but nothing that should require discussion or block merging.

**Comment by SimonJF** [comment]
Fab, thanks! This good to go when CI passes?

**Review by jamescheney — APPROVED** [review]

**Comment by jamescheney** [comment]
huh, I thought I chose "approve" - should be fixed now.

**Comment by SimonJF** [comment]
Wonderful, thanks!

**Pull request merged**
